File: AST.CS/Assembly.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace AST.CS
{
    public class Assembly
    {
        public static bool Preprocessor(string inSourceFile, string includesDir, string srcDir, string objDir)
        {
            return true;
        }
        public static List<AST> Expr(string inProgramExpression)
        {
            return AST.Expr(inProgramExpression);
        }
        public static bool Compile(List<AST> programAST, out List<ASMTree> programASM)
        {
            List <ASMTree> result = new List <ASMTree>();

            foreach (AST node in programAST)
            {

            }

            programASM = result;

            return true;
        }
        public static bool Link(List<ASMTree> program, string outDir, string[] objFiles, out Assembly outAssembly, string outProgramBuildFilenameASM)
        {
            Assembly asm = new Assembly();

            outAssembly = asm;

            return true;
        }
        public static bool CodeGeneratorAssembler(ref Assembly assembly, string outProgramBuildFilenameASM, string objDir)
        {
            return true;
        }
        public static bool AssembleBinaryExecutable(Assembly inAssembly, string objDir, string inProgramBuildFilenameASM, string outProgramBuildFilenameBIN)
        {
            return true;
        }
        public static bool AssembleBinarySharedLibrary(Assembly inAssembly, string objDir, string inProgramBuildFilenameASM, string outProgramBuildFilenameBIN)
        {
            return true;
        }

        public class ASMTree // Derived from AST
        {

        }
        public class ASMConstant
        {

        }
        public class ASMReturn
        {

        }
        public class ASMFunction
        {

        }
        public class ASMFunctionPointer
        {

        }
        public class ASMStruct
        {

        }
        public class ASMIfCondition
        {

        }
        public class ASMForLoop
        {

        }
        public class ASMDoLoop
        {

        }
        public class ASMWhileLoop
        {

        }
        public class ASMProgram
        {

        }
    }
}

File: squash/Assembly.c
#include "Assembly.h"

bool Preprocessor(char* inSourceFile, char* includesDir, char* srcDir, char* objDir)
{
    return true;
}
AST* ExprLexeme(char* inProgramExpression)
{
    return Expr(inProgramExpression);
}
bool Compile(AST* programAST, ASMTree** programASM)
{
    ASMTree* result = (ASMTree*)malloc(sizeof(ASMTree*));

    (*programASM) = result;

    return true;
}
bool Link(ASMTree* program, char* outDir, char* objFiles, Assembly** outAssembly, char* outProgramBuildFilenameASM)
{
    return true;
}
bool CodeGeneratorAssembler(Assembly* assembly, char* outProgramBuildFilenameASM, char** objDir)
{
    return true;
}
bool ILCodeGenerator(Assembly** assembly, char* outProgramBuildFilenameIL)
{
    return true;
}
bool AssembleBinaryExecutable(Assembly* inAssembly, char* objDir, char* inProgramBuildFilenameASM, char* outProgramBuildFilenameBIN)
{
    return true;
}
bool AssembleBinarySharedLibrary(Assembly* inAssembly, char* objDir, char* inProgramBuildFilenameASM, char* outProgramBuildFilenameBIN)
{
    return true;
}
